Using a virtual DOM for REPL view

eliascotto opened the pull request:
@nobody-famous  I tried to replace the webview rendering with an application using React. The rendering is now done with a virtual DOM and the performance are fantastic. The REPL doesn't freeze the webview when it gets a lot of output.

I've run this code without problem (This time 800000). The rendering it's faster than Emacs on my machine.
```
(dotimes (i 800000)
  (format t "~3D " (1+ i))
  (when (= (mod (1+ i) 20) 0)
    (terpri)))
```

To build the React app run `npm run web-watch`

The code here is not finished and refined, but good enough to do some performance testing. I think using a virtual DOM is the key.

nobody-famous commented:
Yeah, that makes sense. Updating the html was probably causing it to reparse the entire thing on every update. Since a virtual DOM works, using js to dynamically add new nodes should also work. I don't have a strong preference, either way. I think it's fairly common now for extensions to use a framework for their web views. If a framework is added, I'll finally have to look into how to bundle the extension. It already yells at me for not doing it.

eliascotto commented:
Ok good, I can finish the PR with some update for the dev documentation and a new build process. Any preference for the build tool? I think [Vite](https://vite.dev/) seems a good candidate.

nobody-famous commented:
This is what it says when I build the package,

```
This extension consists of 660 files, out of which 281 are JavaScript files. For performance reasons, you should bundle your extension: https://aka.ms/vscode-bundle-extension . You should also exclude unnecessary files by adding them to your .vscodeignore: https://aka.ms/vscode-vscodeignore
```
The link points to here https://code.visualstudio.com/api/working-with-extensions/bundling-extension

Packaging is done with a tool called vsce, which you can install with npm if you want to try it. `vsce package` will create the .vsix file that can be installed as an extension. That's what gives the above message.

I'm fine with Vite as long as it conforms to what they want for the bundles, which it should. That would be a step towards moving away from jest, as well, which tends to be a good thing.

Oh, one other thing to try with testing. When a view is hidden, it gets destroyed and needs to be recreated when it's shown again. If you switch to a terminal, for example, and then go back to the repl, the repl view starts from scratch. That's what the `restoreState` stuff is doing.

eliascotto commented:
Thanks for the link I will check it out. Vite uses esbuild under the hood so it would be fine for this I reckon. For tests, `vitests` is Jest compatible so it should be not much of a burden to migrate to.

Yes I'm aware of `onDidChangeVisibility`, I just gave priority to the rendering performance but I'm going to fix it for sure.

eliascotto commented:
Summary of the work done in this PR:

- Wrote a React app in ts to display the output of the REPL and handle messages from the extension as clear, setInput, setOuput, etc
- Adapted `LispRepl.js` webview to render the React app. Added required content-security-policy to the HTML
- Created new bundle for both `extension` and `webview` using ESBuild. I tried with Vite but kept getting errors related to node libraries not available at runtime. Switched to ESBuild for the moment, no performance issues or errors.
- Added Alive version printed at the beginning of REPL to give better context
- Renaming

nobody-famous commented on src/extension.ts:
```diff
@@ -205,6 +205,10 @@ export const activate = async (ctx: Pick<vscode.ExtensionContext, 'subscriptions
     }
+
+    if (state.extension.packageJSON && state.extension.packageJSON.version) {
+        setTimeout(() => ui.addReplOutput(`; Alive REPL (v${state.extension.packageJSON.version})`), 250)
```
This should be moved to the REPL view as initial content. The timeout is arbitrary, so will cause problems depending on the machine. On my slower laptop, the REPL still comes up blank.

nobody-famous commented on src/vscode/Utils.ts:
```diff
@@ -302,3 +302,12 @@ async function updateCompilerDiagnostics(diagnostics: VscodeDiags, notes: Compil
+
+export function getNonce() {
+    let text = '';
+    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
```
I don't think it matters because it's not visible, but this will generate profanity more often than you'd think.

nobody-famous commented on webview/App.tsx:
```diff
@@ -0,0 +1,126 @@
+    useEffect(() => {
+        window.addEventListener('message', messageHandler)
+        return () => window.removeEventListener('message', messageHandler)
+    }, [messageHandler])
```
Need to set up the linting so it'll complain about these lists. If I remove `messageHandler`, it doesn't do anything.  The biggest issue with React is how easily it can go into a render loop because of the hooks, so it needs to check that the dependency lists are accurate.

nobody-famous commented on webview/App.tsx:
```diff
@@ -0,0 +1,126 @@
+import React, { useEffect, useRef, useState } from 'react'
+import { throttle } from './utils'
+
+declare var acquireVsCodeApi: any
```
Should be `const`. Not real happy about the `any`, but there doesn't seem to be a good way around it.

nobody-famous commented on webview/App.tsx:
```diff
@@ -0,0 +1,126 @@
+
+export default function AliveREPL() {
+    const [vscodeApi, setVscodeApi] = useState(null)
+    const scrollReplRef = useRef(null)
+
+    const [packageName, setPackageName] = useState('')
+    const [inputText, setInputText] = useState('')
+    const [outputItems, setOutputItems] = useState([])
+
+    const inputRef = useRef(null)
+    const replOutputRef = useRef(null)
+    const calledOnce = React.useRef(false)
```
Should have the types specified, i.e. `useState<string>('')`

nobody-famous commented on package.json:
```diff
@@ -535,15 +539,19 @@
     "dependencies": {
         "axios": "^1.6.0",
         "node-stream-zip": "^1.15.0",
+        "react": "^19.0.0",
```
Also need dev dependency for `@types/react`.

nobody-famous commented on src/vscode/views/LispRepl.ts:
```diff
@@ -56,25 +60,25 @@ export class LispRepl extends EventEmitter<ReplEvents> implements vscode.Webview
-        webviewView.webview.html = this.getHtmlForView(webviewView.webview)
+        webviewView.webview.html = this.getWebviewContent(webviewView.webview)
+
+        setTimeout(() => {
+            this.setPackage(this.package)
+        }, 200)
```
Similar timeout issue, on slower machines the package initially comes up empty.

nobody-famous reviewed (commented):
I'll look at it more when I get the chance. I found a problem with the LSP that's going to be fun to fix. So far, looks really good, though.

nobody-famous commented on .vscodeignore:
```diff
@@ -2,3 +2,12 @@
+coverage/
+**/*.ts
+!file.ts
+**/*.js.map
```
looks like this also needs to ignore `src` and `node_modules`.

nobody-famous commented on src/vscode/views/LispRepl.ts:
```diff
@@ -125,46 +137,29 @@ export class LispRepl extends EventEmitter<ReplEvents> implements vscode.Webview
-
-                    <script src="${webview.asWebviewUri(jsPath)}"></script>
+                <body>
+                    <div id="root"></div>
```
It works to pass the initial values for package and the version as attributes to the root div and then passing those as props to `App` in `webview/index.tsx`. That might be simpler than trying to do it with message passing.

eliascotto commented:
Thanks for the review. I’ll look into it, I’m on holiday right now.

On Tue, 18 Feb 2025 at 11:51 am, nobody-famous ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/vscode/views/LispRepl.ts
> <https://github.com/nobody-famous/alive/pull/209#discussion_r1958989573>:
>
> > -                            </form>
> -                        </div>
> -                        <div class="repl-input-text-box">
> -                            <div class="repl-input-label" onclick="requestPackage()">
> -                                <span id="repl-package">${this.package}</span>
> -                                >
> -                            </div>
> -                            <form id="repl-input-form" class="repl-input-form" action="">
> -                                <input class="repl-input-text" id="repl-input-text" type="text">
> -                            </form>
> -                        </div>
> -                    </div>
> -
> -                    <script src="${webview.asWebviewUri(jsPath)}"></script>
> +                <body>
> +                    <div id="root"></div>
>
> It works to pass the initial values for package and the version as
> attributes to the root div and then passing those as props to App in
> webview/index.tsx. That might be simpler than trying to do it with
> message passing.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nobody-famous/alive/pull/209#pullrequestreview-2622320850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB5YOLITMSZGFAP6GC5VL6L2QKN33AVCNFSM6AAAAABW52XCWGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMRSGMZDAOBVGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

eliascotto commented:
Updated the PR solving the comments and updating the tests.

nobody-famous commented on src/vscode/views/LispRepl.ts:
```diff
@@ -48,6 +54,8 @@ export class LispRepl extends EventEmitter<ReplEvents> implements vscode.Webview
                         return this.emit('historyDown')
                     case 'userInput':
                         return this.emit('userInput', msg.text ?? '')
+                    case 'viewDidMounted':
```
It doesn't look like this message gets sent from anywhere.

nobody-famous commented on webview/App.tsx:
```diff
@@ -0,0 +1,150 @@
+            if (extensionVersion) {
+                console.log("inside")
+                setOutputItems((prevOutputItems) => {
+                    const versionInfo = `; Alive REPL (v${extensionVersion})`
```
This is duplicated in extension.ts, but it seems like this one might not be getting added to the REPL output list.

nobody-famous reviewed (commented):
I haven't quite figured out why, but switching over to the `Alive Client` output window and back causes the REPL output to change. The version banner always disappears. Clearing the `Alive Client` output window will cause the REPL to completely reset and go back to only the version banner, but not every time. Seems like maybe a race condition? I'm not sure why it's forgetting about it's old content.

eliascotto commented:
Thanks for  testing it! I cleaned out the old duplicated way to print the version information and fixed the disappearing. The problem was that `LispRepl` was trying to restore the old state on `onDidChangeVisibility`, but the App wasn't fully loaded yet.

nobody-famous commented:
This seems to be working fine. Looks like it just needs the tests updated to cover the new stuff and synced with master.

eliascotto commented:
Tests are passing, but for some reason jest is returning `1` instead of `0`. Possibly a [bug](https://github.com/jestjs/jest/issues/9324/) in the library.

```

Test Suites: 22 passed, 22 total
Tests:       408 passed, 408 total
Snapshots:   0 total
Time:        11.413 s
Ran all test suites.
```

eliascotto commented:
@nobody-famous

nobody-famous commented:
Sorry, I haven't been online much. A couple of 7 month old puppies were dumped on my land. The shelters are full so won't take them. I've been trying to find homes for them. Doing that on top of everything else is eating up my personal project time. I haven't written a line of code outside of work in over a week now.

It's failing because of coverage.
```
 LSP.ts                   |     100 |    99.43 |     100 |     100 | 284               
 LispRepl.ts              |   86.53 |       78 |   85.71 |   86.27 | 74-76,83,116-122  
```
The last column are lines that aren't being tested. If you run the tests locally, red numbers means the line is never hit during the tests and yellow means there's a branch that wasn't taken.

eliascotto commented:
No worries at all, that sounds like a tough situation. I appreciate you still taking the time to review this with everything going on. Hope you’re able to find good homes for the pups soon!

I’ve fixed the tests, so coverage should be all good now. Let me know if there’s anything else you'd like me to adjust. Thanks again for the feedback.
